pkg/apis/openstack/validation/infrastructure.go

```diff
@@ -74,66 +78,105 @@ func ValidateInfrastructureConfigUpdate(oldConfig, newConfig *api.Infrastructure
 
 	return allErrs
 }
 
-func validateFloatingPoolNameConstraints(fps []api.FloatingPool, region string, name string, fldPath *field.Path) field.ErrorList {
+func validateFloatingPoolNameConstraints(fps []api.FloatingPool, domain, region string, name string, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	_, errs := findFloatingPool(fps, region, name, fldPath.Child("floatingPoolName"))
+	_, errs := FindFloatingPool(fps, domain, region, name, fldPath.Child("floatingPoolName"))
 	allErrs = append(allErrs, errs...)
 	return allErrs
 }
 
-func findFloatingPool(floatingPools []api.FloatingPool, shootRegion, floatingPoolName string, fldPath *field.Path) (*api.FloatingPool, field.ErrorList) {
+// FindFloatingPool finds best match for given domain, region, and name
+func FindFloatingPool(floatingPools []api.FloatingPool, domain, shootRegion, floatingPoolName string, fldPath *field.Path) (*api.FloatingPool, field.ErrorList) {
 	var (
-		allErrs                = field.ErrorList{}
-		validValues            []string
-		globalCandidate        *api.FloatingPool
-		globalCandidateScore   int
-		regionalCandidate      *api.FloatingPool
-		regionalCandidateScore int
-		regionalFound          bool
+		allErrs               = field.ErrorList{}
+		validValues           = sets.NewString()
+		additionalValidValues = sets.NewString()
 	)
 
-	for _, fp := range floatingPools {
-		// store the first non-regional image for fallback value if no floating pool for the given
-		// region was found
-		if fp.Region == nil && !regionalFound {
-			if match, score := utils.SimpleMatch(fp.Name, floatingPoolName); match {
-				if globalCandidate == nil || globalCandidateScore < score {
-					f := fp
-					globalCandidate = &f
-					globalCandidateScore = score
-				}
+	found := findFloatingPoolCandidate(floatingPools, &domain, &shootRegion, floatingPoolName, false, validValues, additionalValidValues)
+	if found != nil {
+		return found, allErrs
+	}
+
+	nonConstrainingOnly := len(validValues) > 0
+	validValuesRegion := sets.NewString()
+	validValuesDomain := sets.NewString()
+	foundRegion := findFloatingPoolCandidate(floatingPools, nil, &shootRegion, floatingPoolName, nonConstrainingOnly, validValuesRegion, additionalValidValues)
+	foundDomain := findFloatingPoolCandidate(floatingPools, &domain, nil, floatingPoolName, nonConstrainingOnly, validValuesDomain, additionalValidValues)
+	if foundRegion != nil && foundDomain != nil {
+		return foundRegion, allErrs
+	}
+	if foundRegion != nil && len(validValuesDomain) == 0 {
+		return foundRegion, allErrs
+	}
+	if foundDomain != nil && len(validValuesRegion) == 0 {
+		return foundDomain, allErrs
+	}
+	if len(validValuesRegion) != 0 && len(validValuesDomain) != 0 {
+		for value := range validValuesRegion {
+			if validValuesDomain.Has(value) {
+				validValues.Insert(value)
 			}
```

Comment:
IIUC, this can only happen if the same FP is configured twice which is at the same time forbidden by the cloud profile validation logic you added in `pkg/apis/openstack/validation/cloudprofile.go`. 
Seems like unreachable code?
I'm also not sure why we don't allow the union of `validValuesDomain` and `validValuesRegion`. Can you elaborate?

Reply:
This was also my first thought. But the union of `validValuesDomain`and `validValuesRegion` is not quite correct, as constraints of both must be fulfilled.
It must be the intersection of both, but only if both are specified.
The validation logic in the cloud profile only forbids duplicates of the tuple `(name, region, domain)`.

Example:
```yaml
   constraints:
      floatingPools:
      - name: fp1
        region: eu
      - name: fp1
        domain: dom1
      - name: fp2
        region: us
```

Case 1: region `eu`, domain `dom1`
validValuesRegion=`["fp1"]`, validValuesDomain=`["fp1"]` => validValues is the intersection, i.e. `["fp1"]`

Case 2: region `us`, domain `dom1`
validValuesRegion=`["fp2"]`, validValuesDomain=`["fp1"]`,  => validValues is the intersection, i.e. `[]`.

Reply:
I see, thanks for explaining. Can you add a unit test for this case as well?

Also, you can make use `validValuesDomain.Intersection(validValuesRegion)` instead of looping and adding the intersected values separately.

Reply:
This is already fully covered by the "FindFloatingPool table" test (see https://github.com/gardener/gardener-extension-provider-openstack/pull/76/files#diff-aa42ff40241b80a4ca184b047877b3faR460-R462)

Reply:
Don't we miss a case for which the `validValues > 0` after the intersection? At least all tests pass when I add the following panic to `FindFloatingPool`.

	if len(validValuesRegion) != 0 && len(validValuesDomain) != 0 {
		// if region and domain are constrainted separately, only values are valid which are valid for both
		validValues = validValuesRegion.Intersection(validValuesDomain)
		if validValues.Len() > 0 {
			panic("covered")
		}

Reply:
Ok, now also covered in the case an invalid fp is requested (see last commit)